engine/src/main/java/org/hibernate/validator/internal/properties/javabean/JavaBean.java

```diff
@@ -6,32 +6,210 @@
  */
 package org.hibernate.validator.internal.properties.javabean;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Member;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Type;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.stream.Stream;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
 
-import org.hibernate.validator.internal.properties.Property;
 import org.hibernate.validator.internal.properties.ConstrainableType;
-import org.hibernate.validator.internal.util.ReflectionHelper;
+import org.hibernate.validator.internal.util.CollectionHelper;
+import org.hibernate.validator.internal.util.privilegedactions.GetDeclaredConstructors;
+import org.hibernate.validator.internal.util.privilegedactions.GetDeclaredFields;
+import org.hibernate.validator.internal.util.privilegedactions.GetDeclaredMethods;
+import org.hibernate.validator.internal.util.privilegedactions.GetMethods;
+import org.hibernate.validator.spi.properties.ConstrainableExecutable;
+import org.hibernate.validator.spi.properties.GetterPropertySelectionStrategy;
 
 /**
+ * {@link ConstrainableType} that describes a JavaBean.
+ *
  * @author Marko Bekhta
  */
 public class JavaBean implements ConstrainableType {
 
-	private final Class<?> clazz;
+	private final String name;
+	private final List<JavaBeanField> beanFields;
+	private final List<JavaBeanMethod> beanMethods;
+	private final List<JavaBeanGetter> beanGetters;
+	private final List<JavaBeanConstructor> beanConstructors;
+
+	public JavaBean(GetterPropertySelectionStrategy getterPropertySelectionStrategy, Class<?> clazz) {
+		this( getterPropertySelectionStrategy, clazz, false );
+	}
+
+	public JavaBean(GetterPropertySelectionStrategy getterPropertySelectionStrategy, Class<?> clazz, boolean includeNonDeclaredMethods) {
+		this.name = clazz.getName();
```

Comment:
Maybe `includeMethodsFromHierarchy` would be better?

Even in this case, I think I would separate them in another `Map` and have a different method to access them (a method without the `Declared` part as in the reflection API).

Reply:
Ah and I would keep the boolean handy and throw an `IllegalStateException` if the method is called when it being false.

Reply:
Maybe hold on on this particular change as we might not need this feature if we decide to not use this thing for XML metadata.